plugins/filter_aws/aws.c

```diff
@@ -182,190 +227,270 @@ static int cb_aws_init(struct flb_filter_instance *f_ins,
+    /* since get_metadata returned 0, we assume there is at least 1 tag */
+    /* \n is separator, therefore number of items = number of \n + 1 */
+    ctx->tags_count = 1;
+    for (i = 0; i < len; i++) {
+        if (tags_list[i] == '\n') {
+            ctx->tags_count++;
+        }
     }
-    else {
-        flb_plg_debug(ctx->ins, "Using IMDSv1");
+    ctx->tag_keys = flb_calloc(ctx->tags_count, sizeof(flb_sds_t));
+    if (!ctx->tag_keys) {
+        flb_errno();
+        flb_sds_destroy(tags_list);
+        return -1;
     }
-
-    /* Perform request */
-    ret = flb_http_do(client, &b_sent);
-    flb_plg_debug(ctx->ins, "IMDS metadata request http_do=%i, HTTP Status: %i",
-                  ret, client->resp.status);
-
-    if (ret != 0 || client->resp.status != 200) {
-        if (client->resp.payload_size > 0) {
-            flb_plg_debug(ctx->ins, "IMDS metadata request\n%s",
-                          client->resp.payload);
-        }
-        flb_http_client_destroy(client);
-        flb_upstream_conn_release(u_conn);
+    ctx->tag_keys_len = flb_calloc(ctx->tags_count, sizeof(size_t*));
+    if (!ctx->tag_keys_len) {
+        flb_errno();
+        flb_sds_destroy(tags_list);
         return -1;
     }
 
-    if (key != NULL) {
-        /* get the value of the key from payload json string */
-        tmp = flb_json_get_val(client->resp.payload,
-                               client->resp.payload_size, key);
-        if (!tmp) {
-            tmp = flb_sds_create_len("NULL", 4);
-            flb_plg_error(ctx->ins,
-                         "%s is undefined in EC2 instance", key);
+    /* go over the response and initialize tag_keys values */
+    /* code below finds two indices which define tag key and copies them to ctx */
+    while (tag_end <= len) {
+        /* replace \n with \0 to 'clearly' separate tag key strings */
+        if (tags_list[tag_end] == '\n') {
+            tags_list[tag_end] = '\0';
         }
-    } else {
-        tmp = flb_sds_create_len(client->resp.payload, client->resp.payload_size);
+        if ((tags_list[tag_end] == '\0' || tag_end == len) && (tag_start < tag_end)) {
+            /* length of tag key characters is the difference between start and end */
+            /* for instance, if tag name is 'Name\0...', the corresponding values are */
+            /*   tag_start = 0, points to 'N' */
+            /*   tag_end = 4, points to '\0' just after 'e' */
+            /*   f.e.: 4 - 0 = 4, which is equal to len("Name") */
+            tag_key_len = tag_end - tag_start;
+            ctx->tag_keys_len[tag_index] = tag_key_len;
+
+            /* allocate new memory for the tag key value */
+            /* + 1, because we need one more character for \0 */
+            tmp = flb_sds_create_size(tag_key_len + 1);
+            if (!tmp) {
+                flb_errno();
+                flb_sds_destroy(tags_list);
+                return -2;
+            }
+            tmp[tag_key_len] = '\0';
```

Comment:
@PettitWesley I've added this line which should fix your issue.

I assumed that `flb_sds_create_size` returns calloced chunk of memory, but it appears there might be some non-zero characters.

Reply:
Oh yeah it uses malloc so the memory is unitialized: https://github.com/fluent/fluent-bit/blob/master/src/flb_sds.c#L41